flagscale/train/hetero/p2p_communication.py

```diff
@@ -17,10 +17,76 @@
 
 from flagscale.train import get_parallel_context  
-
+from flagscale.train.hetero.parallel_context import ParallelContext
 # Types
 Shape = Union[List[int], torch.Size]
 
+def warm_up_comm_group_hetero(config: ModelParallelConfig):
+    group = None
+    rank = torch.distributed.get_rank()
+    para_ctx = get_parallel_context()
+    pp_groups = para_ctx.get_pipeline_model_parallel_group()
+    tensor_shape = [1, 3]
+    to_send_tensor= torch.empty(
+            tensor_shape,
+            requires_grad=True,
+            device=torch.cuda.current_device(),
+            dtype=config.pipeline_dtype,
+        )
+    to_recv_tensor= torch.empty(
+            tensor_shape,
+            requires_grad=True,
+            device=torch.cuda.current_device(),
+            dtype=config.pipeline_dtype,
+        )
+
+    for pp_group in pp_groups:
+        group_ranks = torch.distributed.get_process_group_ranks(pp_group)
+        if rank == group_ranks[0]:
+            _communicate(
+                tensor_send_next=to_send_tensor,
+                tensor_send_prev=None,
+                recv_prev=False,
+                recv_next=False,
+                tensor_shape=to_recv_tensor.shape,
+                config=config,
+                group=pp_group,
+            )
+        elif rank == group_ranks[-1]:
+            _communicate(
+                tensor_send_next=None,
+                tensor_send_prev=None,
+                recv_prev=True,
+                recv_next=False,
+                tensor_shape=to_recv_tensor.shape,
+                config=config,
+                group=pp_group,
+            )
+        elif rank in group_ranks:
+            _communicate(
+                tensor_send_next=to_send_tensor,
+                tensor_send_prev=None,
+                recv_prev=True,
+                recv_next=False,
+                tensor_shape=to_recv_tensor.shape,
+                config=config,
+                group=pp_group,
+            )
+
+
+def is_inter_mesh_comm(para_ctx: ParallelContext, comm_with_front_layer: bool):
+    assert para_ctx is not None, "Specify ParallelContext Necessary"
+    assert comm_with_front_layer is not None, "Specify Communication Direction Necessary"
+    if comm_with_front_layer:
+        total_prev_pipeline_model_parallel_size = 0
+        for i in range(0, para_ctx._current_process_mesh_index):
+            total_prev_pipeline_model_parallel_size += para_ctx._process_meshes[i]._rank_generator.pp
+        return core.parallel_state.get_pipeline_model_parallel_rank() == total_prev_pipeline_model_parallel_size
+    else:
+        total_current_pipeline_model_parallel_size = 0
+        for i in range(0, min(para_ctx._current_process_mesh_index + 1, len(para_ctx._process_meshes))):
+            total_current_pipeline_model_parallel_size += para_ctx._process_meshes[i]._rank_generator.pp
+        return core.parallel_state.get_pipeline_model_parallel_rank() == total_current_pipeline_model_parallel_size - 1        
```

Comment:
Please add more comments for easier understanding.

Reply:
OK.